### ISSUE TITLE
Assertion failure when debug info, loop metadata and function declaration interact

### ISSUE THREAD

**Issue by aus-intel** [KhronosGroup/SPIRV-LLVM-Translator#524]
Reverse translation of this opencl code
```
float bar(int x);

__kernel void foo(__global float* outPtr, int i) {
#pragma clang loop unroll(enable)
  for (int j = 0; j < i; ++j) {
    outPtr[j] = bar(j);
  }
}
```
fails with assertion failure in debug and segfault in release.
Commands to reproduce:
```
$ clang repro.cl -Xclang -finclude-default-header -c -emit-llvm -O0 -target spir64 -gline-tables-only
$ llvm-spirv repro.bc
$ llvm-spirv -r repro.spv -o repro_rev.bc
```
Result:
```
llvm-spirv: llvm-project/llvm/include/llvm/ADT/ilist_iterator.h:138: llvm::ilist_iterator<OptionsT, IsReverse, IsConst>::reference llvm::ilist_iterator<OptionsT, IsReverse, IsConst>::operator*() const [with OptionsT = llvm::ilist_detail::node_options<llvm::BasicBlock, true, false, void>; bool IsReverse = false; bool IsConst = false; llvm::ilist_iterator<OptionsT, IsReverse, IsConst>::reference = llvm::BasicBlock&]: Assertion `!NodePtr->isKnownSentinel()' failed.
```

**Comment by aus-intel** [comment]
I tried to investigate main source of problem. The issue arises mostly because of this call https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/54f072eb9f1eee8eac799cd701e96c82e9b2d403/lib/SPIRV/SPIRVReader.cpp#L2431 in transFunction.

transLLVMLoopMetadata tries to build LoopInfo for translated function. It keeps member map with metadata. When this map is empty, transLLVMLoopMetadata immediately returns without any attempts to build LoopInfo.
https://github.com/KhronosGroup/SPIRV-LLVM-Translator/blob/54f072eb9f1eee8eac799cd701e96c82e9b2d403/lib/SPIRV/SPIRVReader.cpp#L834-L841
This works when translation goes inorder -- declarations are handled before they are referenced in definitions.

However, when debug info is present, it can forward-reference function definitions. In this case, forward reference is immediately translated as it was real definition point. This leads to out of order translation with very strange effects. In this case it looks like this:
1) Debug info translation is started.
2) Forward reference to foo is met so foo translation is started.
3) Metadata in foo is placed in map
4) Reference to bar is encountered, bar translation started inside of foo translation
5) Bar things processed, entering transLLVMLoopMetadata for bar
6) Map is not empty, translator tries to build LoopInfo for declaration of bar -- failure

**Comment by AlexeySachkov** [comment]
@mlychkov, @AGindinson: FYI, please take a look

**Comment by mlychkov** [comment]
This issue is assigned on @vmaksimo. However, I have no permission to assign it in github.

**Comment by AlexeySachkov** [comment]
Should be resolved by #541. @aus-intel, feel free to re-open if there are still issues

**Comment by aus-intel** [comment]
Failures are gone. I will backport patch to llvm 10 branch.
